File: haskell/glue/src/Glue/Lib/Bool/When.hs
module Glue.Lib.Bool.When where

import Glue.Eval (Eval, eval, throwError)
import Glue.Eval.Exception (wrongArgumentType)
import Glue.IR (IR (..), isTruthy)

when_ :: IR Eval
when_ = Special whenImpl

whenImpl :: [IR Eval] -> Eval (IR Eval)
whenImpl [cond, body] = do
    condVal <- eval cond
    if isTruthy condVal
        then eval body
        else pure Void
whenImpl _ = throwError $ wrongArgumentType ["condition", "body"]

File: dart/glue/lib/src/lib/bool/when.dart
import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

/// When conditional execution special form
/// Mirrors Haskell Glue.Lib.Bool.When.when_ exactly
final Ir when_ = IrSpecial(whenImpl);

/// When conditional execution special form implementation
/// Mirrors Haskell Glue.Lib.Bool.When.whenImpl exactly
Eval<Ir> whenImpl(List<Ir> args) {
  return switch (args) {
    [final cond, final body] => eval(cond).bind((condVal) {
      return isTruthy(condVal) ? eval(body) : Eval.pure(IrVoid());
    }),
    _ => throwError(wrongArgumentType(['condition', 'body'])),
  };
}
